### PR TITLE
Add get_cell() and get_cells() methods to PartialRowData

### PR DESCRIPTION
**Pull request by zakons**
Simple method to get a cell value. If no index supplied, get the most recent value.

### CODE REVIEW

**Comment by zakons** [comment]
Signed the CLA!

**Comment by zakons** [comment]
Added an additional email address for CLA.

**Comment by chemelnucfin** [comment]
Thank you for your contribution.  We require corresponding tests when you change source code.  I was working on something similar and was trying to figure out the unit test but I had the system test working if you want to compare.

Also, @dhermes maybe value would be good enough for the function name?

File: bigtable/google/cloud/bigtable/row_data.py [hunk 1]
```diff
@@ -171,6 +171,24 @@ def row_key(self):
...
+        :param column_family_id: The ID of the column family. Must be of the
+                                 form ``[_a-zA-Z0-9][-_.a-zA-Z0-9]*``.
+        :type column: bytes
+        :param column: The column within the column family where the cell
+                       is located.
+
+        :type index: int
+        :param index: The offset in the series of values, default = 0
+
+        :rtype:  bytes or :class:`int`
```
Comment (chemelnucfin) [inline bigtable/google/cloud/bigtable/row_data.py]: should this be bytes only?

File: bigtable/google/cloud/bigtable/row_data.py [hunk 2]
```diff
@@ -171,6 +171,24 @@ def row_key(self):
         """
         return self._row_key
 
+    def cell_value(self, column_family_id, column, index=0):
+        """Get a cell value
+
+        :type column_family_id: str
+        :param column_family_id: The ID of the column family. Must be of the
+                                 form ``[_a-zA-Z0-9][-_.a-zA-Z0-9]*``.
+        :type column: bytes
```
Comment (chemelnucfin) [inline bigtable/google/cloud/bigtable/row_data.py]: \n here?

File: bigtable/google/cloud/bigtable/row_data.py [hunk 3]
```diff
@@ -171,6 +171,24 @@ def row_key(self):
         """
         return self._row_key
 
+    def cell_value(self, column_family_id, column, index=0):
+        """Get a cell value
+
+        :type column_family_id: str
+        :param column_family_id: The ID of the column family. Must be of the
+                                 form ``[_a-zA-Z0-9][-_.a-zA-Z0-9]*``.
```
Comment (chemelnucfin) [inline bigtable/google/cloud/bigtable/row_data.py]: \n here?

**Comment by chemelnucfin** [comment]
@dhermes I think it LGTM to me after my comments, but I'd like you to look it over.

File: bigtable/google/cloud/bigtable/row_data.py (hunk 1 above)
Comment (zakons) [reply bigtable/google/cloud/bigtable/row_data.py]: This documentation was copied verbatim from method set_cell on the different row classes in row.py.  Any changes here should be reviewed in the context of all the documentation.

Thanks!

Comment (chemelnucfin) [reply bigtable/google/cloud/bigtable/row_data.py]: Hello @zakons, my apologies.  I'm relatively new here and I didn't mean to be offensive.  I wasn't sure of the answers myself and therefore phrased it as a question.  Sorry if it seemed offensive.

File: bigtable/google/cloud/bigtable/row_data.py [hunk 4]
```diff
@@ -171,6 +181,105 @@ def row_key(self):
...
+
+        try:
+            cell = cells[index]
+        except (TypeError, IndexError):
+            num_cells = len(cells)
+            msg = _MISSING_INDEX.format(
+                index, column, column_family_id, num_cells)
+            raise IndexError(msg)
+
+        return copy.deepcopy(cell)
```
Comment (dhermes) [inline bigtable/google/cloud/bigtable/row_data.py]: @lukesneeringer Can you weigh in here on the `deepcopy`? I think I'd rather just return the `Cell` or at least provide a keyword `copy=True` that users can disable?

File: bigtable/google/cloud/bigtable/row_data.py [hunk 5]
```diff
@@ -171,6 +181,105 @@ def row_key(self):
...
+            specified column.
+
+        Raises:
+            KeyError: If ``column_family_id`` is not among the cells stored
+                in this row.
+            KeyError: If ``column`` is not among the cells stored in this row
+                for the given ``column_family_id``.
+        """
+        cells = self._get_cells_no_copy(column_family_id, column)
+        return copy.deepcopy(cells)
```
Comment (dhermes) [inline bigtable/google/cloud/bigtable/row_data.py]: @lukesneeringer Ditto here, I think I'd like to copy the list containing the cells, but not copy the `Cell`-s themselves. WDYT?

File: bigtable/google/cloud/bigtable/row_data.py (hunk 4 above)
Comment (chemelnucfin) [reply bigtable/google/cloud/bigtable/row_data.py]: I thought of having another function to return just the value.

Comment (dhermes) [reply bigtable/google/cloud/bigtable/row_data.py]: The user can just call `cell.value` and I think having two nearly identical methods would be overly "wordy" for the API. (Happy to be overruled though.)

File: bigtable/tests/unit/test_row_data.py [hunk 6]
```diff
@@ -174,6 +177,138 @@ def test_to_dict(self):
...
+        partial_row_data = self._make_one(None)
+        self.assertEqual(partial_row_data._cells, {})
+
+        with self.assertRaises(KeyError) as exc_info:
+            partial_row_data.get_cells(family_name, None)
+
+        expected_arg = row_data._MISSING_COLUMN_FAMILY.format(family_name)
+        self.assertEqual(exc_info.exception.args, (expected_arg,))
+
+    def test_get_cell_bad_column(self):
```
Comment (dhermes) [inline bigtable/tests/unit/test_row_data.py]: Oops, this was a copy-pasta error by me. Needs to be `test_get_cells_bad_column`

File: bigtable/google/cloud/bigtable/row_data.py (hunk 1 above)
Comment (zakons) [reply bigtable/google/cloud/bigtable/row_data.py]: Actually, I agree it should be bytes only, but as I said, I copied the documentation for consistency.  I would encourage it be corrected and made consistent if this is correct. I will follow up on this.

Comment (dhermes) [reply bigtable/google/cloud/bigtable/row_data.py]: @zakons The distinction as that `set_cell()` will accept an integer and then encode the integer as a big-endian `long long` (`>q`) using the [`struct` library][1].

[1]: https://docs.python.org/3/library/struct.html

**Comment by sduskis** [comment]
What do we need to move this forward?

**Comment by zakons** [comment]
I will take some steps over the weekend to move this forward, now that the row iterator is merged.
    On Friday, January 26, 2018, 8:55:03 AM EST, Solomon Duskis <notifications@github.com> wrote:  
 
 
What do we need to move this forward?

**Comment by zakons** [comment]
Superseded by PR #4908.

**Pull request closed without merging**